### PR TITLE
feat: add segments

### PR DESCRIPTION
**Pull request by ykhedher**

### CODE REVIEW

**Comment by ykhedher** [comment]
[Segments frontend](https://trello.com/c/FG54gVeQ/926-segments-frontend)

**Comment by ykhedher** [comment]
[Frontend - Create segment form](https://trello.com/c/WOcP0TIH/941-frontend-create-segment-form)

**Review by FredrikOseberg — CHANGES_REQUESTED** [review]
I haven't had time to look through everything, but here are some superficial functionality issues I found: 
- The searchfield only allows you to select a context field _once_. What if you want to have a constraint two times with different operators?
- You can save a segment even though constraint validation isn't passing
- If you save the constraints, and go back, then forward. All constraints will open in edit mode. 
- If you hit cancel on a new constraint field, it should be removed from the list if it's a new constraint and you haven't saved it yet.

**Comment by FredrikOseberg** [comment]
This text should also be reviewed: 

```
Segment makes it easy for you to define who should be exposed to your feature. The segment is often a collection of constraints and can be reused. Create a segment and use it when you configure an activation strategy on a feature toggle.
```

* A segment is _always_ a collection of constraints, not _often_.

Maybe involve @thomasheartman and Elise/Nicolae here.

**Comment by FredrikOseberg** [comment]
Skjermbilde 2022-03-21 kl  16 21 36

Skjermbilde 2022-03-21 kl  16 21 45

Segments are currently listed twice in the sidebar menu (also validate that this does not render when the segments flag isn't set). And the link takes you to an empty create page instead of the list.

**Comment by FredrikOseberg** [comment]
Skjermbilde 2022-03-21 kl  16 27 31

Should be a connected entity like before, with the stepper flowing "halfway" into the body of the info. But keep it as a single unit.

**Comment by FredrikOseberg** [comment]
Skjermbilde 2022-03-21 kl  16 33 47

Note from UX team: 
* Stepper should have slightly different spacing
* Stepper circles should be adjusted according to sketch and use color scheme from sketch

File: src/component/segments/EditSegment/EditSegment.tsx
Comment (thomasheartman) [inline src/component/segments/EditSegment/EditSegment.tsx:75]: @FredrikOseberg 

I know this is beyond the scope of this PR, but I think we should update the form template component to make the "Learn more" link say something more meaningful:

https://www.a11yproject.com/posts/creating-valid-and-accessible-links/#write-helpful-link-text

A simple solution would be to add "about X" after "Learn more". If it's an issue that the link text gets longer and more unpredictable, we could also hide the extra text from rendering by using a `visually-hidden` class or something, but it'd be nice to have it there.

File: src/component/segments/CreateSegment/CreateSegment.tsx
Comment (thomasheartman) [inline src/component/segments/CreateSegment/CreateSegment.tsx:75]: I'm sure you're aware of this, but the link isn't correct at the moment. We should make sure to update this before it gets merged. Of course, we'll need to write the docs first 😅

File: src/component/segments/CreateSegment/CreateSegment.tsx
Comment (thomasheartman) [inline src/component/segments/CreateSegment/CreateSegment.tsx:71]: The two calls to create and edit template seem awfully similar. If this happens for every feature, is there a way to abstract it in such a way that you don't need to provide the description and links twice? 🤔  Might not be, of course; just seems that it might be easier.

File: src/component/segments/CreateSegment/CreateSegment.tsx
Comment (thomasheartman) [inline src/component/segments/CreateSegment/CreateSegment.tsx:74]: Nice! 😄 I suggest a little rewrite into something like this (prettify as you wish):

```suggestion
            description="Segments make it easy for you to define which of your users should get access to a feature.
            A segment is a reusable collection of constraints.
            You can create and apply a segment when configuring activation strategies for a feature toggle or at any time from the segments page in the navigation menu."
```

**Review by thomasheartman — CHANGES_REQUESTED** [review]
Nice work! 🎉  I've only had a look at the things Fredrik suggested (and little bits around it). Comments, questions, and suggestions are inline. Keep up the good work 🙌🏼

File: src/component/segments/EditSegment/EditSegment.tsx
Comment (olav) [reply src/component/segments/EditSegment/EditSegment.tsx]: Good idea! I'll add an optional custom docs link label and use it in the segment forms, at least.

File: src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx
Comment (thomasheartman) [inline src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx:35]: Is this true only if the constraint is new or is it also true if the constraint has unsaved changes?

File: src/component/common/FormTemplate/FormTemplate.tsx
Comment (thomasheartman) [inline src/component/common/FormTemplate/FormTemplate.tsx:149]: Can we make the link label a required props "Learn more" is never a descriptive link label, so we should always provide more information. It would require changes in a lot of unrelated components, however, so it's probably more at home in a separate PR, but I'd love to hear some thoughts on this.

**Review by thomasheartman — APPROVED** [review]
All the changes I requested have been addressed and none of my new comments are blockers, so I can approve my part 🥳  However, I can't vouch for the rest of the code (too much context I don't have), so I'll leave that to someone else for now.

There's a few inline comments: one's a question (mostly about curiosity, but it means that I didn't understand the code/naming well enough) and one's a request / discussion point that I'd like to hear some thoughts around 😇

File: src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx
Comment (olav) [reply src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx:35]: Right, this is only `true` when the constraint is brand new (i.e. it has never been saved), used by the cancel button.

File: src/component/common/FormTemplate/FormTemplate.tsx
Comment (olav) [reply src/component/common/FormTemplate/FormTemplate.tsx:149]: Yeah, that would be great. A separate PR makes sense to me as well.

File: src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx
Comment (thomasheartman) [reply src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx:35]: Got it! I don't know if it needs a better name or not, it just wasn't clear to me when looking at the diff. Maybe `new`/`isNew` would be better? But again, if it's clear from looking at the code, then that's probably enough. An outside view isn't always helpful for things like this 💁🏼

Comment (olav) [reply src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx:35]: Definitely, `new` seems much better. Will fix.

File: src/component/common/FormTemplate/FormTemplate.tsx
Comment (thomasheartman) [reply src/component/common/FormTemplate/FormTemplate.tsx:149]: Sweet; glad you agree 😄 I went ahead and made a [Trello ticket for it](https://trello.com/c/8qX1K0nC/524-documentation-links-are-not-descriptive)

File: src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx
Comment (FredrikOseberg) [inline src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx:85]: If I'm understanding this correctly, this WeakMap is a key value store where the object is used as the key and the metadata is associated with the object reference? Does it trigger re-renders?

I'm curious about this pattern.

File: src/component/common/FormTemplate/FormTemplate.styles.ts
Comment (FredrikOseberg) [inline src/component/common/FormTemplate/FormTemplate.styles.ts:3]: Should this make it into the theme? There's a discussion regarding how to use the theme properly that I'd like to have at some point.

File: src/component/feature/FeatureStrategy/FeatureStrategyForm/FeatureStrategyForm.tsx
Comment (FredrikOseberg) [inline src/component/feature/FeatureStrategy/FeatureStrategyForm/FeatureStrategyForm.tsx:112]: How is this working on main when flags is an object?

File: src/component/feature/FeatureStrategy/FeatureStrategySegment/FeatureStrategySegmentList.tsx
Comment (FredrikOseberg) [inline src/component/feature/FeatureStrategy/FeatureStrategySegment/FeatureStrategySegmentList.tsx:47]: Could extract this in a variable `lastSegment`, I know this is the expression for the last element of an array - yet I still had to stop and think through what is going on here. Small readability optimisation.

File: src/component/segments/SegmentFormStepList/SegmentFormStepList.tsx
Comment (FredrikOseberg) [inline src/component/segments/SegmentFormStepList/SegmentFormStepList.tsx:31]: Could we optimise readability here? 

```suggestion
                {Array.from({ length: total }).map((_, index) => {
                    const currentStep = index  === current - 1
                    return <FiberManualRecord
                        key={i}
                        className={classNames(
                            styles.circle,
                            currentStep && styles.filledCircle
                        )}
                    />
                })}
```

File: src/component/segments/SegmentFormStepOne/SegmentFormStepOne.tsx
Comment (FredrikOseberg) [inline src/component/segments/SegmentFormStepOne/SegmentFormStepOne.tsx:40]: Could be solved with required prop?

File: src/component/segments/SegmentFormStepTwo/SegmentFormStepTwo.tsx
Comment (FredrikOseberg) [inline src/component/segments/SegmentFormStepTwo/SegmentFormStepTwo.tsx:76]: Context is such a loaded word. I frequently mix it with react context, golang contexts and unleash context. To add to the confusion, there are subtle differences conceptually. Maybe we should rename our internal context references to `CreateUnleashContext` like we have done with the hook?

File: src/component/segments/SegmentList/SegmentList.tsx
Comment (FredrikOseberg) [inline src/component/segments/SegmentList/SegmentList.tsx:169]: Could also use ConditionallyRender here

**Review by FredrikOseberg — COMMENTED** [review]
Looks really good so far! I'll take another look when it's completely ready.

File: src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx
Comment (olav) [reply src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx]: Yeah, just a way to associate extra fields with an object, without extending the object itself. The `useWeakMap` hook takes care of triggering re-renders whenever the map updates.

File: src/component/common/FormTemplate/FormTemplate.styles.ts
Comment (olav) [reply src/component/common/FormTemplate/FormTemplate.styles.ts:3]: Not sure, quite hacky at the moment. Only used by the form steps container, to center it above the main form column.

**Pull request merged**
